linalg/sundials.cpp

```diff
@@ -35,12 +40,392 @@ using namespace std;
+   N_Vector_ID id = MPIPlusX() ? N_VGetVectorID(local_x) : N_VGetVectorID(x);
+#else
+   N_Vector local_x = x;
+   N_Vector_ID id = N_VGetVectorID(x);
+#endif
+
+   // Set the N_Vector data and length from the Vector data and size.
+   switch (id)
+   {
+      case SUNDIALS_NVEC_SERIAL:
+      {
+         MFEM_ASSERT(NV_OWN_DATA_S(local_x) == SUNFALSE, "invalid serial N_Vector");
+         NV_DATA_S(local_x) = HostReadWrite();
+         NV_LENGTH_S(local_x) = size;
+         break;
+      }
+#ifdef MFEM_USE_CUDA
+      case SUNDIALS_NVEC_CUDA:
+      {
+         N_VSetHostArrayPointer_Cuda(HostReadWrite(), local_x);
+         N_VSetDeviceArrayPointer_Cuda(ReadWrite(), local_x);
+         static_cast<N_VectorContent_Cuda>(GET_CONTENT(local_x))->length = size;
+         break;
+      }
+#endif
+#ifdef MFEM_USE_MPI
+      case SUNDIALS_NVEC_PARALLEL:
+      {
+         MFEM_ASSERT(NV_OWN_DATA_P(x) == SUNFALSE, "invalid parallel N_Vector");
+         NV_DATA_P(x) = HostReadWrite();
+         NV_LOCLENGTH_P(x) = size;
+         if (glob_size == 0)
+         {
+            glob_size = GlobalSize();
+
+            if (glob_size == 0 && glob_size != size)
+            {
+               long local_size = size;
+               MPI_Allreduce(&local_size, &glob_size, 1, MPI_LONG,
+                             MPI_SUM, GetComm());
+            }
+         }
+         NV_GLOBLENGTH_P(x) = glob_size;
+         break;
+      }
+#endif
+      default:
+         MFEM_ABORT("N_Vector type " << id << " is not supported");
+   }
+
+#ifdef MFEM_USE_MPI
+   if (MPIPlusX())
+   {
+      if (glob_size == 0)
+      {
+         glob_size = GlobalSize();
+
+         if (glob_size == 0 && glob_size != size)
+         {
+            long local_size = size;
+            MPI_Allreduce(&local_size, &glob_size, 1, MPI_LONG,
+                          MPI_SUM, GetComm());
+         }
+      }
+      static_cast<N_VectorContent_MPIManyVector>(GET_CONTENT(x))->global_length =
+         glob_size;
+   }
+#endif
+}
+
+void SundialsNVector::_SetDataAndSize_()
+{
+#ifdef MFEM_USE_MPI
+   N_Vector local_x = MPIPlusX() ? N_VGetLocalVector_MPIPlusX(x) : x;
+   N_Vector_ID id = MPIPlusX() ? N_VGetVectorID(local_x) : N_VGetVectorID(x);
+#else
+   N_Vector local_x = x;
+   N_Vector_ID id = N_VGetVectorID(x);
+#endif
+
+
+   // The SUNDIALS NVector owns the data if it created it.
+   switch (id)
+   {
+      case SUNDIALS_NVEC_SERIAL:
+      {
+         const bool known = mm.IsKnown(NV_DATA_S(local_x));
+         size = NV_LENGTH_S(local_x);
+         data.Wrap(NV_DATA_S(local_x), size, false);
+         if (known) { data.ClearOwnerFlags(); }
+         break;
+      }
+#ifdef MFEM_USE_CUDA
+      case SUNDIALS_NVEC_CUDA:
+      {
+         double *h_ptr = N_VGetHostArrayPointer_Cuda(local_x);
+         double *d_ptr = N_VGetDeviceArrayPointer_Cuda(local_x);
+         const bool known = mm.IsKnown(h_ptr);
+         size = N_VGetLength_Cuda(local_x);
```

Comment:
Is it possible for `h_ptr` to be known but not `d_ptr` (or vice versa)?

Reply:
I believe only h_ptr can be unknown, but @camierjs can answer definitively.

Reply:
`IsKnown` can only be called with a host pointer according to the documentation, so this must be the case.